Ensure safe RGB data while processing colorbalancergb

jenshannoschwalm opened the pull request:
As reported in #16071 (comments) we can have some visible artefacts in very dark areas.

These are due to negative pixel rgb data in the still linear part of processing. Make sure we don't process those by ensure at least 0 per channel. Both CPU and OpenCL paths have the same issue.

I think this is correct and also safe, @flannelhead @ralfbrown would appreciate your comments on this.

ralfbrown commented:
Seems OK.  Clipping away negatives will of course cause changed output, but that is preferable to having the following operation blow up due to negative values....

jenshannoschwalm commented:
> Seems OK. Clipping away negatives will of course cause changed output, but that is preferable to having the following operation blow up due to negative values....

Yes and we already clip negatives at the input stage of the module ...

ralfbrown commented:
May also be related to #14796.

TurboGit commented:
Nice, this fixes color issues in dark shadows like reported in #14796.

TurboGit commented:
From integration test:

Without this PR:

![image](https://github.com/darktable-org/darktable/assets/467069/4f82c6f2-9a83-4af3-9b36-20119a60c069)


With this PR:

![image](https://github.com/darktable-org/darktable/assets/467069/3baf41cd-57f2-41ad-a205-f702251785b1)

jenshannoschwalm commented:
> this fixes color issues in dark shadows like reported in #14796.

Indeed and i am pretty sure some OpenCL issues with this module are fixed also.

TurboGit commented:
One regression 0083, but looking at this the new output is better to me. The diff is not huge but still big:

```
Test 0083-colorbalancergb
      Image mire1.cr2
      CPU & GPU version differ by 28788 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 1.32945
      Avg dE                   : 0.00502
      Std dE                   : 0.05284
      ----------------------------------
      Pixels below avg + 0 std : 98.97 %
      Pixels below avg + 1 std : 98.97 %
      Pixels below avg + 3 std : 99.02 %
      Pixels below avg + 6 std : 99.14 %
      Pixels below avg + 9 std : 99.53 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 9.91327
      Avg dE                   : 0.39864
      Std dE                   : 1.27543
      ----------------------------------
      Pixels below avg + 0 std : 86.35 %
      Pixels below avg + 1 std : 92.01 %
      Pixels below avg + 3 std : 95.93 %
      Pixels below avg + 6 std : 99.93 %
      Pixels below avg + 9 std : 100.00 %
      ----------------------------------
      Pixels above tolerance   : 6.72 %
 
  FAILS: image visually changed
         see diff.png for visual difference
         (395199 pixels changed)
```

TurboGit commented on src/iop/colorbalancergb.c:
```diff
@@ -808,7 +808,7 @@ void process(struct dt_iop_module_t *self,
     dt_vector_powf(scaled_RGB, midtones, RGB);
     for_each_channel(c)
-      RGB[c] = RGB[c] * sign[c] * d->white_fulcrum;
+      RGB[c] = MAX(0.0f, RGB[c] * sign[c] * d->white_fulcrum);
```
I have a doubt though, what about the sign? As noted for the OpenCL case we really do want to keep the sign:

`// midtones : power with sign preservation`

The `fmaxf()` does break this.

TurboGit reviewed (commented):
A doubt...

TurboGit commented on data/kernels/extended.cl:
```diff
@@ -877,7 +877,7 @@ colorbalancergb (read_only image2d_t in, write_only image2d_t out,
 
   // midtones : power with sign preservation
-  RGB = sign(RGB) * native_powr(fabs(RGB) / white_fulcrum, midtones) * white_fulcrum;
+  RGB = fmax(0.0f, sign(RGB) * native_powr(fabs(RGB) / white_fulcrum, midtones) * white_fulcrum);
```
And indeed, this seems equivalent to removing `sign(RGB)` as we take fabs(RGB) anyway.

TurboGit replied:
Tested and the following patch also fixes the issue:

```
diff --git a/data/kernels/extended.cl b/data/kernels/extended.cl
index 7ce955268b..273d274d91 100644
--- a/data/kernels/extended.cl
+++ b/data/kernels/extended.cl
@@ -877,7 +877,7 @@ colorbalancergb (read_only image2d_t in, write_only image2d_t out,
   // factorization of : (RGB[c] * (1.f - alpha) + RGB[c] * d->shadows[c] * alpha) * (1.f - beta)  + RGB[c] * d->highlights[c] * beta;
 
   // midtones : power with sign preservation
-  RGB = sign(RGB) * native_powr(fabs(RGB) / white_fulcrum, midtones) * white_fulcrum;
+  RGB = native_powr(fabs(RGB) / white_fulcrum, midtones) * white_fulcrum;
 
   // for the non-linear ops we need to go in Yrg again because RGB doesn't preserve color
   LMS = gradingRGB_to_LMS(RGB);

```

But is that correct? I mean the code and comment about preserving the sign really mean that this was wanter :)

jenshannoschwalm replied:
Let me investigate source in more depth. We do some sort of sign preserving gamma correction here so very neg values in RGB result in neg values too.

jenshannoschwalm replied:
1. I now understand why there were no artefacts in the JzAzBz code, there the conversion via `dt_XYZ_2_JzAzBz` takes care
2. we can & should do the "input-sanity check" later in the specific code for UCS before `dt_XYZ_to_xyY`
3. Surprise,  surprise - spotted a possible div by zero in this code too. Also in OpenCL explaining why some drivers don't like this module. (BTW i checked afterwards, we still need the input sanity check) Also checked the JzAzBz code does not have this problem.

TurboGit commented on data/kernels/extended.cl:
```diff
@@ -991,8 +992,8 @@ colorbalancergb (read_only image2d_t in, write_only image2d_t out,
 
-    float P = HCB.y;
-    float W = sin_T * HCB.y + cos_T * HCB.z;
+    const float P = fmax(1e-9f, HCB.y);
```
Would it makes sense to use `FLT_MIN` instead of `1e-9f`?

TurboGit commented on src/iop/colorbalancergb.c:
```diff
@@ -922,8 +923,8 @@ void process(struct dt_iop_module_t *self,
 
-      float P = HCB[1];
-      float W = sin_T * HCB[1] + cos_T * HCB[2];
+      const float P = MAX(1e-9, HCB[1]); // as HCB[1] is at least zero we don't fiddle with sign
```
Likewise.

TurboGit reviewed (commented):
Just a comment. Otherwise it works fine on my side and no diff on 0083 test, only a small diff on 0093 which is barely visible when looking at the two images.

jenshannoschwalm commented:
Fixed by using FLT_MIN in both cases, checked the OpenCL standard and for 1.2 it's safe to use this too, shouldn't be vendor depending. squashed and force-pushed.

jenshannoschwalm commented:
Release note:
Fix garbeld output and possibly OpenCL errors for very dark regions in colorbalance rgb,

TurboGit reviewed (approved):
All good to me now, thanks!

jenshannoschwalm commented:
Wouldn't this be also for 4.6.1 ?

TurboGit commented:
Sure, will be merged soon.

TurboGit commented:
Now merged in 4.6.x branch.
